main/api/src/mill/api/PathRef.scala

```diff
@@ -37,6 +39,37 @@ case class PathRef private (
 
 object PathRef {
 
+  /**
+   * This class maintains a cache of already validated paths.
+   * It is threadsafe and meant to be shared between threads, e.g. in a ThreadLocal.
+   */
+  class ValidatedPaths() {
+    private val map = new ConcurrentHashMap[Int, PathRef]()
+
+    /**
+     * Revalidates the given [[PathRef]], if required.
+     * It will only revalidate a [[PathRef]] if it's value for [[PathRef.revalidate]] says so and also considers the previously revalidated paths.
+     * @throws PathRefValidationException If a the [[PathRef]] needs revalidation which fails
+     */
+    def revalidateIfNeededOrThrow(pathRef: PathRef): Unit = {
+      def mapKey(pr: PathRef): Int = (pr.path, pr.quick, pr.sig).hashCode()
+      pathRef.revalidate match {
+        case Revalidate.Never => // ok
+        case Revalidate.Once if map.contains(mapKey(pathRef)) => // ok
+        case _ =>
```

Comment:
Should this line be `case Revalidate.Always`? That would make it more clear that this is a fixed set of options, and there isn't some unbounded number of cases that the `case _` is meant to catch

Reply:
Yeah, it's currently `case Revalidate.Once | Revalidate.Always`. But I can also make it explicit, so we get a incomplete match warning if we extend the cases.